plutus-core/plutus-core/src/PlutusCore/Builtin/Emitter.hs

```diff
@@ -1,27 +1,43 @@
 {-# LANGUAGE RankNTypes #-}
 
 module PlutusCore.Builtin.Emitter
-    ( Emitter (..)
-    , emitM
+    ( ChurchList (..)
+    , Emitter (..)
+    , runEmitter
+    , emit
     ) where
 
+import Control.Monad.Writer.Strict
 import Data.Text (Text)
 
--- | A monad for logging that does not hardcode any concrete first-order encoding and instead packs
--- a @Monad m@ constraint and a @Text -> m ()@ argument internally, so that built-in functions that
--- do logging can work in any monad (for example, @CkM@ or @CekM@), for which there exists a
--- logging function.
-newtype Emitter a = Emitter
-    { unEmitter :: forall m. Monad m => (Text -> m ()) -> m a
-    } deriving (Functor)
+newtype ChurchList a = ChurchList
+    { unChurchList :: forall r. (a -> r -> r) -> r -> r
+    }
+
+instance Semigroup (ChurchList a) where
+    ChurchList k1 <> ChurchList k2 = ChurchList $ \f -> k1 f . k2 f
+    {-# INLINE (<>) #-}
+
+instance Monoid (ChurchList a) where
+    mempty = ChurchList $ \_ -> id
+    {-# INLINE mempty #-}
 
--- newtype-deriving doesn't work with 'Emitter'.
-instance Applicative Emitter where
-    pure x = Emitter $ \_ -> pure x
-    Emitter f <*> Emitter a = Emitter $ \emit -> f emit <*> a emit
+instance Foldable ChurchList where
+    foldr f z (ChurchList k) = k f z
+    {-# INLINE foldr #-}
+
+singletonChurchList :: a -> ChurchList a
+singletonChurchList x = ChurchList ($ x)
+{-# INLINE singletonChurchList #-}
+
+newtype Emitter a = Emitter
+    { unEmitter :: Writer (ChurchList Text) a
+    } deriving newtype (Functor, Applicative, Monad)
 
-instance Monad Emitter where
-    Emitter a >>= f = Emitter $ \emit -> a emit >>= \x -> unEmitter (f x) emit
+runEmitter :: Emitter a -> (a, ChurchList Text)
+runEmitter = runWriter . unEmitter
+{-# INLINE runEmitter #-}
 
-emitM :: Text -> Emitter ()
-emitM text = Emitter ($ text)
+emit :: Text -> Emitter ()
```

Comment:
I managed to get myself worried here about whether you could build up massive lists of strings at low cost.  I suppose you could only do that inside a builtin though, so it shouldn't be an issue. It doesn't matter if someone uses `trace` many times for example because the contents will be dumped each time it returns.